src/algebra/algebra/basic.lean

```diff
@@ -1053,244 +1056,229 @@ by rw [algebra_compatible_smul A r (a • m), smul_smul, algebra.commutes, mul_s
   f (r • m) = r • f m :=
 by rw [algebra_compatible_smul A r m, linear_map.map_smul, ←algebra_compatible_smul A r (f m)]
 
-instance : has_coe (M →ₗ[A] N) (M →ₗ[R] N) :=
-⟨λ f, ⟨f.to_fun, λ x y, f.map_add' x y, λ r n, map_smul_eq_smul_map _ _ _⟩⟩
+namespace linear_map
 
-end is_scalar_tower
+variables (R) {A M N}
 
-section restrict_scalars
-/- In this section, we describe restriction of scalars: if `S` is an algebra over `R`, then
-`S`-modules are also `R`-modules. -/
+/-- The `R`-linear map induced by an `A`-linear map when `A` is an algebra over `R`. -/
+def restrict_scalars (f : M →ₗ[A] N) : M →ₗ[R] N :=
+{ to_fun := f,
+  map_add' := λ x y, f.map_add x y,
+  map_smul' := λ c x, map_smul_eq_smul_map _ _ _ }
 
-section semimodule
+variables (R A M N)
 
-variables (R : Type*) [comm_semiring R] (S : Type*) [semiring S] [algebra R S]
-variables (E : Type*) [add_comm_monoid E] [semimodule S E]
-variables {F : Type*} [add_comm_monoid F] [semimodule S F]
+instance coe_is_scalar_tower : has_coe (M →ₗ[A] N) (M →ₗ[R] N) :=
+⟨restrict_scalars R⟩
+
+variables (R) {A M N}
+
+@[simp, norm_cast squash] lemma coe_restrict_scalars_eq_coe (f : M →ₗ[A] N) :
+  (f.restrict_scalars R : M → N) = f := rfl
+
+@[simp, norm_cast squash] lemma coe_coe_is_scalar_tower (f : M →ₗ[A] N) :
+  ((f : M →ₗ[R] N) : M → N) = f := rfl
+
+/-- `A`-linearly coerce a `R`-linear map from `M` to `A` to a function, given an algebra `A` over
+a commutative semiring `R` and `M` a semimodule over `R`. -/
+def lto_fun (R : Type u) (M : Type v) (A : Type w)
+  [comm_semiring R] [add_comm_monoid M] [semimodule R M] [comm_ring A] [algebra R A] :
+  (M →ₗ[R] A) →ₗ[A] (M → A) :=
+{ to_fun := linear_map.to_fun,
+  map_add' := λ f g, rfl,
+  map_smul' := λ c f, rfl }
 
 /--
-When `E` is a module over a ring `S`, and `S` is an algebra over `R`, then `E` inherits a
-module structure over `R`, called `module.restrict_scalars' R S E`.
-We do not register this as an instance as `S` can not be inferred.
+For `r : R`, and `f : M →ₗ[A] N` (where `A` is an `R`-algebra) we define
+`(r • f) m = f (r • m)`.
 -/
-def semimodule.restrict_scalars' : semimodule R E :=
-{ smul      := λ c x, (algebra_map R S c) • x,
-  one_smul  := by simp,
-  mul_smul  := by simp [mul_smul],
-  smul_add  := by simp [smul_add],
-  smul_zero := by simp [smul_zero],
-  add_smul  := by simp [add_smul],
-  zero_smul := by simp [zero_smul] }
+instance algebra_has_scalar : has_scalar R (M →ₗ[A] N) :=
```

Comment:
Wouldn't this instance conflict with `has_scalar R (M →ₗ[R] N)`, since we have `algebra R R` and `is_scalar_tower.left : is_scalar_tower R R M`?

Reply:
One can delete the `has_scalar R (M →ₗ[R] N)` instance then.

Reply:
I've changed the def to be defeq to the old one. (And an example that checks this.)